File: src/sim/simulationEngine.ts
import { assertCanonicalInterval, nextGameWeekBoundary, type CanonicalTimestamp } from "./canonicalTime";
import type { RngState } from "./deterministicRng";

export type SimulationEvent = {
  at: CanonicalTimestamp;
  type: string;
  detail?: string;
};

export type SimulationEnvelope<TWorld> = {
  startedAt: CanonicalTimestamp;
  lastSimulatedAt: CanonicalTimestamp;
  rng: RngState;
  world: TWorld;
};

export type SimulationStepResult<TWorld> = {
  world: TWorld;
  rng: RngState;
  events?: SimulationEvent[];
};

export type SimulationContext = {
  from: CanonicalTimestamp;
  to: CanonicalTimestamp;
  rng: RngState;
  startedAt: CanonicalTimestamp;
};

export type SimulationAdapter<TWorld> = {
  /**
   * Return the next canonical milestone strictly after `after` and at or before `to`.
   * Construction completion, technician arrival/repair completion, operating-block settlement and
   * similar real-time jobs belong here. `startedAt` is the canonical epoch so domain adapters can
   * map game-week/day/hour coordinates to absolute timestamps without consulting wall-clock time.
   */
  nextMilestoneAt?(
    world: TWorld,
    after: CanonicalTimestamp,
    to: CanonicalTimestamp,
    startedAt: CanonicalTimestamp,
  ): CanonicalTimestamp | undefined;

  /** Resolve ordinary simulation for [from, to). UI/rendering must not participate. */
  advanceInterval(world: TWorld, context: SimulationContext): SimulationStepResult<TWorld>;

  /** Resolve all canonical milestone changes at one timestamp before a weekly settlement. */
  resolveMilestonesAt?(world: TWorld, at: CanonicalTimestamp, rng: RngState, startedAt: CanonicalTimestamp): SimulationStepResult<TWorld>;

  /** Resolve the end-of-game-week settlement/report boundary. */
  resolveGameWeekBoundary?(world: TWorld, at: CanonicalTimestamp, rng: RngState, startedAt: CanonicalTimestamp): SimulationStepResult<TWorld>;
};

export type AdvanceSimulationResult<TWorld> = {
  envelope: SimulationEnvelope<TWorld>;
  events: SimulationEvent[];
};

function appendEvents(target: SimulationEvent[], events: SimulationEvent[] | undefined) {
  if (events?.length) target.push(...events);
}

export function advanceSimulation<TWorld>(
  input: SimulationEnvelope<TWorld>,
  to: CanonicalTimestamp,
  adapter: SimulationAdapter<TWorld>,
): AdvanceSimulationResult<TWorld> {
  assertCanonicalInterval(input.lastSimulatedAt, to);
  if (input.lastSimulatedAt === to) return { envelope: input, events: [] };

  let cursor = input.lastSimulatedAt;
  let world = input.world;
  let rng = input.rng;
  const events: SimulationEvent[] = [];

  while (cursor < to) {
    const weekBoundary = nextGameWeekBoundary(input.startedAt, cursor);
    const milestone = adapter.nextMilestoneAt?.(world, cursor, to, input.startedAt);
    if (milestone !== undefined && (milestone <= cursor || milestone > to)) {
      throw new Error("Simulation adapter returned an invalid milestone timestamp.");
    }

    const next = Math.min(to, weekBoundary, milestone ?? Number.POSITIVE_INFINITY);

    if (next > cursor) {
      const advanced = adapter.advanceInterval(world, { from: cursor, to: next, rng, startedAt: input.startedAt });
      world = advanced.world;
      rng = advanced.rng;
      appendEvents(events, advanced.events);
      cursor = next;
    }

    if (milestone !== undefined && cursor === milestone) {
      const resolved = adapter.resolveMilestonesAt?.(world, cursor, rng, input.startedAt);
      if (resolved) {
        world = resolved.world;
        rng = resolved.rng;
        appendEvents(events, resolved.events);
      }
    }

    if (cursor === weekBoundary) {
      const resolved = adapter.resolveGameWeekBoundary?.(world, cursor, rng, input.startedAt);
      if (resolved) {
        world = resolved.world;
        rng = resolved.rng;
        appendEvents(events, resolved.events);
      }
    }
  }

  return {
    envelope: {
      ...input,
      lastSimulatedAt: to,
      world,
      rng,
    },
    events,
  };
}

File: native/SaunaCore/Sources/SaunaCore/SimulationEngine.swift
import Foundation

public struct SimulationEvent: Codable, Equatable, Sendable {
    public var at: Int64
    public var type: String
    public var detail: String?

    public init(at: Int64, type: String, detail: String? = nil) {
        self.at = at
        self.type = type
        self.detail = detail
    }
}

public struct SimulationEnvelope<World: Codable & Equatable & Sendable>: Codable, Equatable, Sendable {
    public var startedAt: Int64
    public var lastSimulatedAt: Int64
    public var rng: RNGState
    public var world: World

    public init(startedAt: Int64, lastSimulatedAt: Int64, rng: RNGState, world: World) {
        self.startedAt = startedAt
        self.lastSimulatedAt = lastSimulatedAt
        self.rng = rng
        self.world = world
    }
}

public struct SimulationContext: Sendable {
    public let from: Int64
    public let to: Int64
    public let rng: RNGState
    public let startedAt: Int64
}

public struct SimulationStepResult<World: Codable & Equatable & Sendable>: Sendable {
    public var world: World
    public var rng: RNGState
    public var events: [SimulationEvent]

    public init(world: World, rng: RNGState, events: [SimulationEvent] = []) {
        self.world = world
        self.rng = rng
        self.events = events
    }
}

public struct SimulationAdapter<World: Codable & Equatable & Sendable>: Sendable {
    public var nextMilestoneAt: @Sendable (World, Int64, Int64, Int64) -> Int64?
    public var advanceInterval: @Sendable (World, SimulationContext) -> SimulationStepResult<World>
    public var resolveMilestonesAt: @Sendable (World, Int64, RNGState, Int64) -> SimulationStepResult<World>
    public var resolveGameWeekBoundary: @Sendable (World, Int64, RNGState, Int64) -> SimulationStepResult<World>

    public init(
        nextMilestoneAt: @escaping @Sendable (World, Int64, Int64, Int64) -> Int64? = { _, _, _, _ in nil },
        advanceInterval: @escaping @Sendable (World, SimulationContext) -> SimulationStepResult<World>,
        resolveMilestonesAt: @escaping @Sendable (World, Int64, RNGState, Int64) -> SimulationStepResult<World> = { world, _, rng, _ in
            SimulationStepResult(world: world, rng: rng)
        },
        resolveGameWeekBoundary: @escaping @Sendable (World, Int64, RNGState, Int64) -> SimulationStepResult<World> = { world, _, rng, _ in
            SimulationStepResult(world: world, rng: rng)
        }
    ) {
        self.nextMilestoneAt = nextMilestoneAt
        self.advanceInterval = advanceInterval
        self.resolveMilestonesAt = resolveMilestonesAt
        self.resolveGameWeekBoundary = resolveGameWeekBoundary
    }
}

public struct AdvanceSimulationResult<World: Codable & Equatable & Sendable>: Sendable {
    public var envelope: SimulationEnvelope<World>
    public var events: [SimulationEvent]
}

public enum SimulationEngineError: Error, Equatable {
    case invalidMilestone
}

public enum SimulationEngine {
    public static func advance<World: Codable & Equatable & Sendable>(
        _ input: SimulationEnvelope<World>,
        to: Int64,
        adapter: SimulationAdapter<World>
    ) throws -> AdvanceSimulationResult<World> {
        try CanonicalTime.assertInterval(from: input.lastSimulatedAt, to: to)
        if input.lastSimulatedAt == to {
            return AdvanceSimulationResult(envelope: input, events: [])
        }

        var cursor = input.lastSimulatedAt
        var world = input.world
        var rng = input.rng
        var events: [SimulationEvent] = []

        while cursor < to {
            let weekBoundary = CanonicalTime.nextGameWeekBoundary(origin: input.startedAt, after: cursor)
            let milestone = adapter.nextMilestoneAt(world, cursor, to, input.startedAt)
            if let milestone, (milestone <= cursor || milestone > to) {
                throw SimulationEngineError.invalidMilestone
            }

            let next = min(to, weekBoundary, milestone ?? Int64.max)
            if next > cursor {
                let advanced = adapter.advanceInterval(
                    world,
                    SimulationContext(from: cursor, to: next, rng: rng, startedAt: input.startedAt)
                )
                world = advanced.world
                rng = advanced.rng
                events.append(contentsOf: advanced.events)
                cursor = next
            }

            if let milestone, cursor == milestone {
                let resolved = adapter.resolveMilestonesAt(world, cursor, rng, input.startedAt)
                world = resolved.world
                rng = resolved.rng
                events.append(contentsOf: resolved.events)
            }

            if cursor == weekBoundary {
                let resolved = adapter.resolveGameWeekBoundary(world, cursor, rng, input.startedAt)
                world = resolved.world
                rng = resolved.rng
                events.append(contentsOf: resolved.events)
            }
        }

        return AdvanceSimulationResult(
            envelope: SimulationEnvelope(
                startedAt: input.startedAt,
                lastSimulatedAt: to,
                rng: rng,
                world: world
            ),
            events: events
        )
    }
}
